### PR TITLE
RDoc-2163-Option to PreventETL from deleting filtered documents - …

### PR DESCRIPTION
**Pull request by serezhub**
…moved doc from 5.1 to 5.3
Ignore the first commit

### CODE REVIEW

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 1]
```diff
@@ -443,36 +451,59 @@ function loadTimeSeriesOfUsersBehavior(doc, ts)
 In order to define deletion handling for a specific collection use the following signature:
 
 {CODE-BLOCK:javascript}
-function deleteDocumentsOf<CollectionName>Behavior(docId) {
+function deleteDocumentsOf<CollectionName>Behavior(docId, deleted) {
```
Comment (arekpalinski) [inline Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:454]: Maybe let's show both signatures:

`function deleteDocumentsOf<CollectionName>Behavior(docId)`

and 

`function deleteDocumentsOf<CollectionName>Behavior(docId, deleted)`

?

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 2]
```diff
@@ -443,36 +451,59 @@ function loadTimeSeriesOfUsersBehavior(doc, ts)
...
-   - `<CollectionName>` needs to be substituted by a real collection name that ETL script is working on (same convention as for `loadTo` method)
-   - The first parameter is the identifier of a deleted document.
+- `<CollectionName>` needs to be substituted by a real collection name that the ETL script is working on (same convention as for [loadTo](../../../server/ongoing-tasks/etl/raven#transformation-script-options) 
+  method)
+   - e.g. `function deleteDocumentsOfOrdersBehavior(docId, deleted) {return [true | false];}`  
+
+| Parameter | Type | Description |
+| - | - | - |
+| **docId** | `string` | The identifier of a deleted document. |
+| **deleted** | `bool` | Optional and therefore doesn't affect existing code. If you don't include the `deleted` parameter, RavenDB will execute the function without checking if the document was deleted from the source database. <br/> If you include `deleted`, RavenDB will check if the document was indeed deleted or just updated.  If the document was deleted in the source database, it will also delete the document in the destination database. |
```
Comment (arekpalinski) [inline Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:466]: I think we are missing explicit explanation that we're doing deletes also when we sent an update of a document. And this parameter allows to distinguish between those two cases.

The original motivation for that was that we had an issue that we were propagating the deletion of a document which was already sent to the destination, then it was updated and it filtered by the condition in the script.

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 3]
```diff
@@ -443,36 +451,59 @@ function loadTimeSeriesOfUsersBehavior(doc, ts)
...
 {NOTE/}
 
 {NOTE: Generic function for deletion handling}
 
-Another option is the usage of generic function for deletion handling:
+Another option is the usage of generic function for collection specific deletion handling:
 
 {CODE-BLOCK:javascript}
-function deleteDocumentsBehavior(docId, collection) {
+function deleteDocumentsBehavior(docId, collection, deleted) {
```
Comment (arekpalinski) [inline Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:480]: Same as above. Let's show both signatures

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 1 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:454]: Done

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 3 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:480]: Done

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 2 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown]: Now I explained it in the intro to this section.

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 3 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:535]: done

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 1 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:489]: done

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 2 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown]: I just want to be clear about the motivation for this project.  
A doc in the destination with a different collection gets deleted by default every time the source is updated or deleted.  
We want to 
1. Enable people to save the outdated/older versions instead of deleting them if they want to.
2. Make sure that they understand the logic of the ID changing with each ETL update when the collection is different.
3. Let them choose whether they want to delete the older version when the source is deleted, and to choose if they want to delete it when the source is simply updated, but not deleted...which is why we added the parameter `deleted` that checks if the source doc was deleted.
4. Am I missing anything?

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 4]
```diff
@@ -448,14 +459,34 @@ function loadTimeSeriesOfUsersBehavior(doc, ts)
 
 {NOTE: Collection specific function deletion handling}
 
-In order to define deletion handling for a specific collection use the following signature:
+* ETL is triggered any time a document is updated or deleted in a collection where ETL is set.
```
Comment (arekpalinski) [inline Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:477]: It's triggered also on _put_ of a document

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 5]
```diff
@@ -448,14 +459,34 @@ function loadTimeSeriesOfUsersBehavior(doc, ts)
 
 {NOTE: Collection specific function deletion handling}
 
-In order to define deletion handling for a specific collection use the following signature:
+* ETL is triggered any time a document is updated or deleted in a collection where ETL is set. 
+
+* To define deletion handling for a specific collection use the following signature 
+  that checks if the document was deleted or updated in the source database.
```
Comment (arekpalinski) [inline Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:480]: I wouldn't say that the signature itself "checks" that. A user can do the check based on the parameter provided to the method.

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 6]
```diff
@@ -448,14 +459,34 @@ function loadTimeSeriesOfUsersBehavior(doc, ts)
 
 {NOTE: Collection specific function deletion handling}
 
-In order to define deletion handling for a specific collection use the following signature:
+* ETL is triggered any time a document is updated or deleted in a collection where ETL is set. 
+
+* To define deletion handling for a specific collection use the following signature 
+  that checks if the document was deleted or updated in the source database. 
+  If it was deleted in the source and the `return` is set to `true`,
```
Comment (arekpalinski) [inline Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:481]: "return` is set to `true`," ?? I'm not sure what you meant here but that is not correct for sure

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 7]
```diff
@@ -448,14 +459,34 @@ function loadTimeSeriesOfUsersBehavior(doc, ts)
...
 
-In order to define deletion handling for a specific collection use the following signature:
+* ETL is triggered any time a document is updated or deleted in a collection where ETL is set. 
+
+* To define deletion handling for a specific collection use the following signature 
+  that checks if the document was deleted or updated in the source database. 
+  If it was deleted in the source and the `return` is set to `true`, 
+  it will also be deleted in the destination without loading a replacement. 
+
+* If the `return` is set to `false`, or if the document was not deleted,
```
Comment (arekpalinski) [inline Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:484]: "OR if the document was not deleted," - this is not how it works, there is no "OR". We provide the values of parameters to the behavior function and _a user_ decides if it should be deleted or not. He does it by returning `true / false`.

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 8]
```diff
@@ -448,14 +459,34 @@ function loadTimeSeriesOfUsersBehavior(doc, ts)
...
+* If the `return` is set to `false`, or if the document was not deleted, 
+  the document in the destination will not be deleted.
 
 {CODE-BLOCK:javascript}
 function deleteDocumentsOf<CollectionName>Behavior(docId, deleted) {
    return [true | false]; 
 }
 {CODE-BLOCK/}
 
+Alternately, the following signature will not check if the document was indeed deleted.
```
Comment (arekpalinski) [inline Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown]: Again, the signature doesn't check itself. The user can do the check

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 9]
```diff
@@ -474,14 +505,25 @@ function deleteDocumentsOf<CollectionName>Behavior(docId, deleted) {
 
 {NOTE: Generic function for deletion handling}
 
-Another option is the usage of generic function for collection specific deletion handling:
+Another option is the usage of a generic function for collection-specific deletion handling.
+
+The following signature will use the `deleted` parameter to check whether a document was deleted in the source database.
```
Comment (arekpalinski) [inline Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:531]: Same here - the signature doesn't "use" the parameter. The user can use it to make a decision and decide what to return from the deletion behavior function

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 2 above)
Comment (arekpalinski) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown]: This is okay

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 4 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:477]: added `created, updated...`

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 5 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:480]: I moved the description into the code sample comments to make it easier to understand. 
```
function deleteDocumentsOf<CollectionName>Behavior(docId, deleted) {
    // If any document in the specified source collection is modified but is not deleted,  
    // then the ETL will not send a delete command to the destination.  
    if (deleted == false) 
    return [false]; 
    // If the source document was deleted, the destination will also be deleted and there will be no replacement.
    else return true;
```

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 6 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:481]: The comments are above the code lines now, which is cleaner than explaining them in a paragraph before the sample.

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 7 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:484]: done.

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 8 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown]: changed it

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 9 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:531]: Changed this whole section to: 
```
* This means that if `return` is `true` and  
  if a source document was updated, but not deleted,  
  it will be deleted in the destination before the updated document with a new ID is loaded to replace the previous one.  
   * If the source document was deleted, the destination will also be deleted and there will be no replacement.
* If `return` is `false`, a 'historical' set of the document will accumulate in the destination collection every time 
  the source document is updated.  
  The number at the end of the ID will automatically increment with every new version. 
   * If the document was deleted from the source, the set of old versions of the document will remain in the destination 
     because the delete behavior returns `false`.
```

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 10]
```diff
@@ -447,7 +447,7 @@ function loadTimeSeriesOfUsersBehavior(doc, ts)
      in which the number at the end is incremented with each version.  
       * e.g. `"...profile/0000000000000000019-B"` will become `".../profile/0000000000000000020-B"`
       * The word before the number is the collection name and the letter after the number is the node.  
-        In this case, I am looking at a document in the collection "Profile" which is in a database in node "B" 
+        In this case, I am looking at two versions of a document in the collection "profile" which is in a database in node "B"
```
Comment (arekpalinski) [inline Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:450]: Why lowercased? Typically the collection names starts with big letter. If user defined `loadToProfile()` then the collection name will be `Profile` just the doc it will be lowercased

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 11]
```diff
@@ -479,17 +481,16 @@ function loadTimeSeriesOfUsersBehavior(doc, ts)
...
-
 To define deletion handling when the source and destination collections are the same, use the following sample. 
 
 {CODE-BLOCK:javascript}
 function deleteDocumentsOf<CollectionName>Behavior(docId, deleted) {
     // If any document in the specified source collection is modified but is not deleted,  
-    // then the ETL will not send a delete command to the destination.  
+    // then the ETL will not send a delete command to the destination, thus saving a history of it.  
     if (deleted == false) 
     return [false];
```
Comment (arekpalinski) [inline Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:491]: Minor - formatting issue

```
if (deleted == false) 
    return [false]; 
```

instead of 
```
if (deleted == false) 
return [false]; 
```

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 12]
```diff
@@ -474,34 +479,36 @@ function loadTimeSeriesOfUsersBehavior(doc, ts)
...
-
-Leaving out the `delete` parameter will trigger the command regardless of whether the source document was deleted. 
-This means that if a source document was updated, but not deleted, 
-it will be deleted in the destination before the updated document is loaded to replace it if `return` is set to `true`.  
+Alternately, leaving out the `deleted` parameter will not allow you to check if the source document was deleted and will 
+trigger the command regardless of whether the source document was deleted.  
 
-If it is set to `false`, a 'historical' list of the documents will accumulate in the destination collection every time 
-the source document is updated.  
+* This means that if `return` is `true` and
```
Comment (arekpalinski) [inline Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:501]: Instead of saying "if `return` is `true`" I'd rather rephrase it to "if the method returns `true`". It's more natural since `return` is a keyword in programming languages and we don't set the value to `return` - it's not a variable. We just return the value from the function

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 13]
```diff
@@ -528,16 +535,32 @@ function deleteDocumentsOf<CollectionName>Behavior(docId) {
...
+To define deletion handling when the source collection may be different from the destination collection , use the following sample where: 
 
 {CODE-BLOCK:javascript}
 function deleteDocumentsBehavior(docId, collection, deleted) {
-   return [true | false]; 
+    // If the destination collection is different than the source and  
+    // if any document in the specified source collection is modified but is not deleted,  
+    // then the ETL will not send a delete command to the destination (the old document versions will remain)
+    if (deleted == false) 
+    return false;
```
Comment (arekpalinski) [inline Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:547]: Formatting 

```
if (deleted == false) 
    return false; 
```

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 10 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:450]: I see..the doc ID is a filename, so all of the letters are lowercase.
Done

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 11 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:491]: Good catch.  I thought I took out the [ ]...  I took it out now.

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 12 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:501]: I see.  Done

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 13 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:547]: Done

**Review by arekpalinski — APPROVED** [review]

**Review by reebhub — CHANGES_REQUESTED** [review]

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 14]
```diff
@@ -0,0 +1,698 @@
...
+
+{PANEL/}
+
+{PANEL: Revisions}
+
+* Revisions are _not_ sent by the ETL process.  
+  But, if revisions are configured on the destination database, then when the target document is overwritten by the ETL process a revision will be created as expected.  
+{PANEL/}
+
+{PANEL: Deletions}
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:453]: Done.  
I changed it to 
```Upon source document modifications, ETL is set by default to delete and replace the destination documents.  

If you want to control the way deletions are handled in the destination database, 
you can change the default settings with the configurable functions described in this section.

* [Why documents are deleted by default]...
```

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 15]
```diff
@@ -0,0 +1,698 @@
...
+  in which the number at the end is incremented with each version.  
+  e.g. `"...profile/0000000000000000019-B"` will become `".../profile/0000000000000000020-B"`
+  The word before the number is the collection name and the letter after the number is the node.  
+  In this case, the document's collection is "Profile", which is in a database in node "B", 
+  and which has been updated via ETL 20 times.
+
+* If the ETL is defined to load the documents to more than one collection, 
+  by default it will delete, and if it's not deleted in the source, it will replace all of the documents with the same prefix.
+   * For example: 
+     1. Document `employees/1-A` is processed by ETL and put into the `People` and `Sales` collections with IDs:
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown]: Done

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 16]
```diff
@@ -0,0 +1,698 @@
...
+
+* If the ETL is defined to load the documents to more than one collection, 
+  by default it will delete, and if it's not deleted in the source, it will replace all of the documents with the same prefix.
+   * For example: 
+     1. Document `employees/1-A` is processed by ETL and put into the `People` and `Sales` collections with IDs:  
+        `employees/1-A/people/0000000000000000001-A` and `employees/1-A/sales/0000000000000000001-A`. 
+     2. Deletion or modification of the `employees/1-A` document on the source side triggers sending a command that deletes **all** documents 
+        having the following prefix in their ID: `employees/1-A`.  
+
+Deletions can be controlled by defining deletion behavior functions in the ETL script.
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:500]: Odd. That's not how it looks on my end...
image

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 17]
```diff
@@ -0,0 +1,704 @@
...
+| **docId** | `string` | The identifier of a deleted document. |
+| **collection** | `string` | The name of a collection. |
+| **deleted** | `bool` | Optional and therefore doesn't affect existing code. If you don't include the `deleted` parameter, RavenDB will execute the function without checking if the document was deleted from the source database. <br/> If you include `deleted`, RavenDB will check if the document was indeed deleted or just updated.  If you set `return true` and the document was deleted in the source database, it will also delete the document in the destination database.  |
+
+| Return | Description |
+| - | - |
+| **bool** | If the returned value is `true`, the document will be deleted. |
+
+
+## ETL script with deletion behavior defined
```
Comment (reebhub) [inline Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:651]: start with "Deletions: .." / "Deletions Example: .."

**Review by reebhub — CHANGES_REQUESTED** [review]

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 18]
```diff
@@ -0,0 +1,704 @@
...
+
+If we load a document to the same collection as the source, then the ID is preserved and no special approach is needed. Deletion in the source results in 
+sending a single delete command in the destination for a given ID.  
+  
+But, when we load documents to different collections, then the [ID](../../../server/ongoing-tasks/etl/raven#documents-identifiers) is different.  
+The source isn't aware of the new IDs created. This forces us to load new documents with incremented IDs instead of overwriting the fields in existing documents.  
+
+* Each updated version of the document gets a [server generated ID](../../../client-api/document-identifiers/working-with-document-identifiers#server-side-generated-ids)
+  in which the number at the end is incremented with each version.  
+  * For example:
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown]: done

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown (hunk 17 above)
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown]: done

File: Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown [hunk 19]
```diff
@@ -0,0 +1,704 @@
...
+{PANEL/}
+
+{PANEL: Deletions}
+
+Upon source document modifications, ETL is set by default to delete and replace the destination documents.  
+
+If you want to control the way deletions are handled in the destination database, 
+you can change the default settings with the configurable functions described in this section.
+
+* [Why documents are deleted by default](../../../server/ongoing-tasks/etl/raven#why-documents-are-deleted-by-default-in-the-destination-database)
```
[the remark this answers is not in the document]
Comment (serezhub) [reply Documentation/5.3/Raven.Documentation.Pages/server/ongoing-tasks/etl/raven.markdown:460]: They all work again.

**Review by reebhub — APPROVED** [review]
